Add example view components as a gem `shared_view_components` (#949)
ViewComponent/view_component: docs/guide/creating_a_gem.md

---
layout: default

> title: Creating a Gem

Reviewer: Do you think this document should go under Guide? Or maybe a new parent heading?

Reviewer 2: What about putting it under _Practical Examples_ as introduced in #946?

Reviewer: @Spone I'm torn about adding that new section. Maybe we need to take step back and think about our docs organization more broadly.

Reviewer 2: Feel free to reach out if you'd like to talk about it (or open a new issue for that?)

Reviewer: It might be nice to have a group chat about this with @coderberry and whoever else might be interested. I know that he had some thoughts about higher-level docs goals.

---

# Creating a Gem

Start by creating a new Rails engine with the following arguments:

```sh
rails plugin new shared-view_components \
  --full \
  --skip-namespace \
  --skip-action-mailer \
  --skip-action-mailbox \
  --skip-action-text \
  --skip-active-record \
  --skip-active-job \
  --skip-active-storage \
  --skip-action-cable \
  --skip-turbolinks \
  --skip-jbuilder \
  --skip-gemfile-entry # only use this when adding a gem into an existing Rails app
```

Update the gemspec. Add the dependency:

```ruby
spec.add_dependency "view_component", ">= 2.32"
```

> Afterwards, add the require command into `shared/view_components.rb`:

Reviewer: ```suggestion
Add a `require` to `shared/view_components.rb`:
```

```ruby
require "view_component/engine"
```

> Let's create our first component:

Reviewer: ```suggestion
To create the first component:
```

```sh
✗ bin/rails g component Shared::Example title --skip-namespace
      create  app/components/shared/example_component.rb
      invoke  test_unit
      create    test/components/shared/example_component_test.rb
      invoke  erb
      create    app/components/shared/example_component.html.erb
```

> Update `example_component.html.erb` to the following:

Reviewer: ```suggestion
Update `example_component.html.erb` to:
```

```erb
<span title="<%= @title %>"><%= content %></span>
```

> Now, let't write a test for the component. Open `test/components/example_component_test.rb`

Reviewer: What do you think about using headings for each step? 

I could see having a `Testing` heading here.

```ruby
require "test_helper"

class ExampleComponentTest < ViewComponent::TestCase
  def test_renders
    assert_equal(
      %(<span title="Hello">World!</span>),
      render_inline(Shared::ExampleComponent.new(title: "Hello").with_content("World!")).css("span").to_html
    )
  end
end
```

Run the test:

```sh
✗ bin/rails test
Run options: --seed 33407

# Running:

..

Finished in 0.047348s, 42.2404 runs/s, 42.2404 assertions/s.
2 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

> Great! Now lets add test the gem in the dummy test app. To start, we need to add `require "view_component/test_helpers"` to `test_helper.rb`

Reviewer: I think this line is not grammatically correct.

Also, might we use a term besides `dummy`? Some might find it less than inclusive ❤️

Reviewer 2: We went for `sandbox` in https://github.com/github/view_component/pull/985#issuecomment-870604847

Reviewer 2: I just noticed that the `dummy` name comes from the `test/dummy` folder created by the `rails plugin new` generator. Changing the term here would mean that this folder has to be renamed by the user after running `rails plugin new`. Not very practical, I guess. Maybe you should advocate for changing the term in Rails?

```ruby
# Configure Rails Environment
ENV["RAILS_ENV"] = "test"

require_relative "../test/dummy/config/environment"
require "rails/test_help"

require "view_component/test_helpers"
```

> Now create a controller where we can test the component:

Reviewer: ```suggestion
Now create a controller using the component:
```

```sh
cd test/dummy
bin/rails g controller examples index
      create  app/controllers/examples_controller.rb
       route  get 'examples/index'
      invoke  erb
      create    app/views/examples
      create    app/views/examples/index.html.erb
      invoke  test_unit
      create    test/controllers/examples_controller_test.rb
      invoke  helper
      create    app/helpers/examples_helper.rb
      invoke    test_unit
      invoke  assets
      invoke    css
      create      app/assets/stylesheets/examples.css
```

> Now let's add the Shared::ExampleComponent to `test/dummy/app/views/examples/index.html.erb`:

Reviewer: ```suggestion
Now add the Shared::ExampleComponent to `test/dummy/app/views/examples/index.html.erb`:
```

```erb
<%= render(Shared::ExampleComponent.new(title: "Hello")) { "World!" } %>
```

> Now let's add a test to `navigation_test.rb` to ensure the component is rendering properly:

Reviewer: ```suggestion
Then add a test to `navigation_test.rb` to ensure the component is rendering properly:
```

```ruby
require "test_helper"

class NavigationTest < ActionDispatch::IntegrationTest
  test "Shared::ExampleComponent renders correctly" do
    get "/examples/index"
    assert_equal 200, status
    assert response.parsed_body.include?('<span title="Hello">World!</span>')
  end
end
```

> Finally, run `rails/test` again to ensure it's working properly.

Reviewer: Can you replace `it's` with direct reference? I like to avoid unclear antecedents when possible: https://writingcommons.org/article/identifying-and-addressing-unclear-pronouns-antecedents/
